### PR TITLE
Remove Protobuf from the ABT SDK

### PR DESCRIPTION
**Pull request by christibbs**
- Removes the generated `ABTExperimentPayload` from the ABT SDK.
- Replaces it with an NSObject subclass that contains the same metadata.
- Fixes up unit tests.

- Updates Remote Config and In-App Messaging to consume "new" API

### CODE REVIEW

File: FirebaseABTesting.podspec [hunk 1]
```diff
@@ -45,10 +45,10 @@ Firebase Cloud Messaging and Firebase Remote Config in your app.
     'HEADER_SEARCH_PATHS' => '"${PODS_TARGET_SRCROOT}"'
   }
   s.dependency 'FirebaseCore', '~> 6.8'
```
Comment (paulb777) [inline FirebaseABTesting.podspec:47]: Delete `GPB_USE_PROTOBUF_FRAMEWORK_IMPORTS=1 ` above.

File: scripts/check_imports.swift [hunk 2]
```diff
@@ -99,7 +99,9 @@ func checkFile(_ file: String, isPublic: Bool) {
       let importFile = line.components(separatedBy: " ")[1]
       if inSwiftPackageElse {
         if importFile.first != "<" {
-          genError("Import error: \(file):\(lineNum) Import in SWIFT_PACKAGE #else should start with \"<\".")
```
Comment (paulb777) [inline scripts/check_imports.swift:102]: Looks like your local swiftformat might be the wrong version - see https://github.com/firebase/firebase-ios-sdk#code-formatting or just revert this change.

File: FirebaseABTesting/Sources/Public/ABTExperimentPayload.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  ABTExperimentPayloadExperimentOverflowPolicyUnrecognizedValue = 999,
+  ABTExperimentPayloadExperimentOverflowPolicyUnspecified = 0,
+  ABTExperimentPayloadExperimentOverflowPolicyDiscardOldest = 1,
+  ABTExperimentPayloadExperimentOverflowPolicyIgnoreNewest = 2,
+};
+
+@interface ABTExperimentLite : NSObject
+@property(nonatomic, readonly, copy) NSString *experimentId;
+
+- (instancetype)initWithExperimentId:(NSString *)experimentId;
```
Comment (chliangGoogle) [inline FirebaseABTesting/Sources/Public/ABTExperimentPayload.h:29]: maybe mark this designated initializer?

File: FirebaseABTesting/Sources/ABTExperimentPayload.m [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    _triggerEvent = dictionary[kExperimentPayloadKeyTriggerEvent];
+    _setEventToLog = dictionary[kExperimentPayloadKeySetEventToLog];
+    _activateEventToLog = dictionary[kExperimentPayloadKeyActivateEventToLog];
+    _clearEventToLog = dictionary[kExperimentPayloadKeyClearEventToLog];
+    _timeoutEventToLog = dictionary[kExperimentPayloadKeyTimeoutEventToLog];
+    _ttlExpiryEventToLog = dictionary[kExperimentPayloadKeyTTLExpiryEventToLog];
+
+    // Other.
+    NSDate *experimentStartTime = [[[self class] experimentStartTimeFormatter]
+        dateFromString:dictionary[kExperimentPayloadKeyExperimentStartTime]];
```
Comment (chliangGoogle) [inline FirebaseABTesting/Sources/ABTExperimentPayload.m:82]: nit, just want to make sure if dictionary[kExperimentPayloadKeyExperimentStartTime] is nil, it won't crash here.

File: FirebaseABTesting/Sources/FIRExperimentController.m [hunk 5]
```diff
@@ -34,19 +35,19 @@
...
 
 /// Deserialize the experiment payloads.
 ABTExperimentPayload *ABTDeserializeExperimentPayload(NSData *payload) {
+  // Verify that we have a JSON object.
   NSError *error;
-  ABTExperimentPayload *experimentPayload = [ABTExperimentPayload parseFromData:payload
-                                                                          error:&error];
-  if (error) {
+  id JSONObject = [NSJSONSerialization JSONObjectWithData:payload options:kNilOptions error:&error];
+  if (JSONObject == nil) {
```
Comment (chliangGoogle) [inline FirebaseABTesting/Sources/FIRExperimentController.m:46]: When there's no experiment running, is it possible that JSON object is just nil without error?

File: FirebaseABTesting.podspec (hunk 1 above)
Comment (christibbs) [reply FirebaseABTesting.podspec:46]: Done in dade3e0.

File: FirebaseABTesting/Sources/FIRExperimentController.m (hunk 5 above)
Comment (christibbs) [reply FirebaseABTesting/Sources/FIRExperimentController.m:46]: Since this method gets called from an enumeration, the payload data should never be nil.  Only way this serialization should fail is if it's a non-JSON object.

File: FirebaseABTesting/Sources/ABTExperimentPayload.m (hunk 4 above)
Comment (christibbs) [reply FirebaseABTesting/Sources/ABTExperimentPayload.m]: The changes in c2b6de3 prevent this.

File: FirebaseABTesting/Sources/Public/ABTExperimentPayload.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 Google
```
Comment (paulb777) [inline FirebaseABTesting/Sources/Public/ABTExperimentPayload.h:1]: If we're not expanding the public API, this file should be put in Sources/Private and added to s.private_header_files in the podspec.

Comment (christibbs) [reply FirebaseABTesting/Sources/Public/ABTExperimentPayload.h]: Done in 3542804.

File: FirebaseABTesting/CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,6 @@
+# v3.4.0
+- [changed] Removed Protbuf dependency (#5890).
```
Comment (paulb777) [inline FirebaseABTesting/CHANGELOG.md:2]: nit: Protobuf spelling

File: FirebaseABTesting/Sources/ABTExperimentPayload.m [hunk 8]
```diff
@@ -0,0 +1,151 @@
+// Copyright 2020 Google
```
Comment (paulb777) [inline FirebaseABTesting/Sources/ABTExperimentPayload.m:1]: Copyright 2020 Google LLC

New copyrights should follow go/copyright

File: FirebaseABTesting.podspec [hunk 9]
```diff
@@ -35,20 +35,19 @@ Firebase Cloud Messaging and Firebase Remote Config in your app.
    'FirebaseCore/Sources/Private/*.h',
   ]
   s.requires_arc = base_dir + '*.m'
-  s.public_header_files = base_dir + 'Public/*.h', base_dir + 'Protos/developers/mobile/abt/proto/*.h'
-  s.private_header_files = base_dir + 'Protos/developers/mobile/abt/proto/*.h'
+  s.public_header_files = base_dir + 'Public/*.h'
```
Comment (paulb777) [inline FirebaseABTesting.podspec:38]: It's weird, but the Private headers should be in the `public_header_files` line additionally.  Full details in the podspec docs.

Comment (christibbs) [reply FirebaseABTesting.podspec]: Fixed in 8e836ca.

File: FirebaseABTesting/CHANGELOG.md (hunk 7 above)
Comment (christibbs) [reply FirebaseABTesting/CHANGELOG.md]: Fixed in c75c41b.

File: FirebaseABTesting/Sources/ABTExperimentPayload.m (hunk 8 above)
Comment (christibbs) [reply FirebaseABTesting/Sources/ABTExperimentPayload.m]: Fixed in 511e0b3.

**Comment by paulb777** [comment]
The travis issues reproduce locally with Xcode 10.3 targeting an iPhone5 simulator. Likely there is a 32bit/64bit data processing issue.

Screen Shot 2020-06-26 at 9 45 18 AM

File: FirebaseABTesting/Sources/Public/FIRExperimentController.h [hunk 10]
```diff
@@ -17,15 +17,15 @@
 @class ABTExperimentPayload;
 
 // Forward declaration to avoid importing into the module header
-typedef NS_ENUM(int32_t, ABTExperimentPayload_ExperimentOverflowPolicy);
```
Comment (paulb777) [inline FirebaseABTesting/Sources/Public/FIRExperimentController.h:20]: Changing the type name will break RemoteConfig users who lock their version and do a `pod update`. Can we keep the type name unchanged until the next major version.

Otherwise, we could consider bumping the major version since we don't really have a public API for ABT.

cc: @ryanwilson

Comment (christibbs) [reply FirebaseABTesting/Sources/Public/FIRExperimentController.h:20]: I'd prefer bumping the major version since the underscores aren't idiomatic Objective-C. What do you guys think?

**Comment by christibbs** [comment]
> The travis issues reproduce locally with Xcode 10.3 targeting an iPhone5 simulator. Likely there is a 32bit/64bit data processing issue.
> 
> Screen Shot 2020-06-26 at 9 45 18 AM

Thanks for looking into that. I didn't have iPhone 5 on my simulators list. I changed the millisecond properties to `long long` in de42dd0.

**Comment by paulb777** [comment]
I'm not sure about changing the double floating types to int types. I would have thought it would be changing the intValue's to int64Value's - https://developer.apple.com/documentation/foundation/nsnumber/1416870-int64value.

@ryanwilson Will you answer the API question? Are we ok to bump the major version of FirebaseABTesting without a Firebase major version since there are no public ABTesting APIs?

**Comment by ryanwilson** [comment]
After considering it, I think I'm fine with bumping the major version of ABTesting without a Firebase major version.

**Comment by christibbs** [comment]
> After considering it, I think I'm fine with bumping the major version of ABTesting without a Firebase major version.

Great. Bumped major version in 84d865f.

**Comment by christibbs** [comment]
> One more change and then LGTM:
> 
> Bump the FirebaseABTesting version in Firebase.podspec

Done in 8e68e3e.
